### ISSUE TITLE
Single unified UI for OrganicMaps

### ISSUE THREAD

**Issue by pktiuk** [organicmaps/organicmaps#9407]
**Is your feature request related to a problem? Please describe.**

With this raw idea I want to tackle several issues:

- Linux desktop interface seems to be lagging behind (does not look very well, is not convergent, does not work well with touch and in general is not very familiar for people using it on mobile devices).    
- There is no online/web version of organic maps and creating it would require also creating new UI again
- Developing several UIs in parallel seems to generate a lot of overhead for developers
- Developer willing to add some UI features (for example like adding new button for changing building color) to this app needs to setup several development environments and needs to test it several times (setup env on linux, later install, android studio, and later buy a mac and setup environment for iOS)
- This repository uses multiple languages for different modules: Python, C++, Java, Swift, which makes reviewing and development difficult. 

**Describe the ideal solution**

There is a single convergent and universal frontend for this entire app working with all (or at least most) of target platforms.  
Currently, there are a lot of technologies and libraries which could be used for this purpose. Some example technologies:

- QML - part of QT library, which can target many types of targets
 - Flutter or flutter-based bindings like [flet](https://flet.dev/) (linked issue: https://github.com/organicmaps/organicmaps/issues/9292 )
 - [Kivy](https://kivy.org/)
 - [Compose Multiplatform](https://github.com/JetBrains/compose-multiplatform) - framework developed by JetBrains
- Xamarin
- [Dioxus](https://github.com/DioxusLabs/dioxus)

Selecting one of them is a big decission, which should be taken by Organic Maps dev team

**Describe alternatives you have considered**

- Use web technologies and create simple Electron based app
- Create a regular desktop app which could be compiled into WebAssembly and somehow package it for multiple platforms
- Changing just a framework for desktop UI for something mode cross-platorm (but without mobile support) like  [iced](https://github.com/iced-rs/iced)

**Additional context**

I just wanted to contribute something to the place editor UI and I saw, that it would require a lot of changes in multiple clients. So I gave up.

This issue could result in implementing another, more universal frontend which could be developed concurrently until it would be mature and could replace other frontends one by one.

Labels for this issue I find relevant: Desktop, Android, iOS, DevEx, Raw Idea, UI, UX

**Comment by RedAuburn** [comment]
it's true that feature parity isn't perfect, but it's not too bad between iOS & Android (the main platforms at the moment). Flutter apps etc always feel a bit strange IMO.

**Comment by pktiuk** [comment]
> Flutter apps etc always feel a bit strange IMO

I don't think this has to be flutter. Flutter is just one of technologies which can be used for this purpose.

> feature parity isn't perfect, but it's not too bad

This is not about current feature parity, but general maintenance burden linked with ensuring this feature parity. (And general development of multiple versions of app at once.)

**Comment by AndrewShkrob** [comment]
![No](https://c.tenor.com/kbQv5cIBq8IAAAAd/tenor.gif)

First, considering the amount of time and effort required to rewrite the whole project using a new technology stack that none of our active contributors know on the necessary level, I'd say it's impossible.
Secondly, all these cross-platform things that are developed to save your time would just waste your time on:
1. learning them
2. fighting with weird bugs that you can't find on the internet (because only you use this tool)
3. fighting with bugs that even developers of this tool won't know what's wrong because it's reproducible only on some old devices with incorrectly working custom ROMs (hello, Android 😏)

Also, they are very slow and heavy (compared to native implementation). 
<img width="739" alt="image" src="https://github.com/user-attachments/assets/9e350578-cf85-4bc0-a49c-139cb95b9bf1">

And based on my experience (with QtQuick and Xamarin) it's just a waste of time.
Here is an example of how I wasted a week of my life trying to replace Qt.Widgets with QtQuick for our desktop app (because it is "better", "easier" and all other good adjectives they use in ads)

https://github.com/user-attachments/assets/6277c5de-eabb-4d16-9946-a87847fb499e

And this is just a "weird bug". Believe me, the performance was much worse :)

These tools are good for prototyping when you have a week or a month to present a simple app (or even just a feature) on every platform. But when it comes to a huge complex application that requires good performance and memory/battery efficiency cross-platform tools are not your choice.

**Comment by AndrewShkrob** [comment]
> Linux desktop interface seems to be lagging behind (does not look very well, is not convergent, does not work well with touch and in general is not very familiar for people using it on mobile devices).

The desktop app was developed just for development purposes. Our focus is on mobile platforms. Yes, it's good to have maps on your laptop but be realistic, you would open it just a few times a year. We have some users that have mobile phones or tablets with linux. I don't have statistics but I believe it's not even 0.1% of all our users. Spending time on a desktop app is not feasible.

> There is no online/web version of organic maps and creating it would require also creating new UI again

There is no reason to have an online app for offline maps. Just imagine you open a site and it tells you to download 200 Mb to use it. This could be useful only when OM starts to support online maps. This won't happen in the near future.

> Developing several UIs in parallel seems to generate a lot of overhead for developers

Look at it from the other side. You don't need to develop and test all platforms simultaneously. You need to develop a UI for only one platform. It's much easier and faster. Yes, that's not convenient for the project but it is for the contributor. You don't need to know specific cross-platform tools, you don't need to know all platforms you're developing for. You need to know only one platform. And yes, when you use cross-platform tools you have to know every platform you're developing for. Because every time there is something in your cross-platform code that is not working on one specific platform. This is my experience.

> * Developer willing to add some UI features (for example like adding new button for changing building color) to this app needs to setup several development environments and needs to test it several times (setup env on linux, later install, android studio, and later buy a mac and setup environment for iOS)
> * This repository uses multiple languages for different modules: Python, C++, Java, Swift, which makes reviewing and development difficult.

Unfortunately, that's how the real world works. Even using a cross-platform tool, building and testing applications for macOS/iOS is only possible on macOS :)

**Comment by AndrewShkrob** [comment]
And can you name at least 2-3 <ins>real</ins> apps written on f.e. Flutter that are more complex than a calculator? It would be perfect if they work with native code. Ideally rendering - OpenGL, Vulkan. I believe you won't find any.

**Comment by pktiuk** [comment]
> 1. learning them

Yes, that's a valid point, but the same could be said about current stack of OrganicMaps which requires C++, Python, Swift, Java and Objective C.  It does not make life easier for new contributors willing to contribute something.

>  2. fighting with weird bugs that you can't find on the internet (because only you use this tool)
> 3. fighting with bugs that even developers of this tool won't know what's wrong because it's reproducible only on some old devices with incorrectly working custom ROMs (hello, Android 😏)

Why do you assume, that you have to choose some unpopular and niche technology/framework. I just came up with some examples to start discussion. Before going into technical details I would prefer to at least rule out some solutions and focus on the most promising.

> There is no reason to have an online app for offline maps. Just imagine you open a site and it tells you to download 200 Mb to use it.

Why not use the logic of an offline map app to create online maps?    
You presented one way to implement this. Store data and do processing on the client's side (requires big donwloads), but I thought about making these kinds of operations on the backend and just visualize result for the client (just like online apps are supposed to work). But I think this is not the place for discussing possible implementations of it.

> The desktop app was developed just for development purposes. Our focus is on mobile platforms.

This is a valid point and makes sense. Need for this kind of desktop app is not comparable with mobile use case.

> Look at it from the other side. You don't need to develop and test all platforms simultaneously.

But instead of testing on 2-3 platforms, you have to reimplement the same feature several times if you want to have feature parity.

> And can you name at least 2-3 real apps written on f.e. Flutter that are more complex than a calculator? It would be perfect if they work with native code. Ideally rendering - OpenGL, Vulkan. I believe **you won't find any**.


I think you should check Google Earth. There is an interesting story about reimplementing their GUI to Flutter https://medium.com/flutter/extreme-ui-adaptability-in-flutter-how-google-earth-supports-every-use-case-on-earth-6db4661e7a17

You could also check an app with complex logic and native rendering called PUBG, which incorporated flutter as a part of their frontend to avoid rewriting UI many times [link to short dev story](https://www.youtube.com/watch?v=03ETVuWZiBc) 

Are these 2 examples complex and `real` enough?

**Comment by AndrewShkrob** [comment]
> Why do you assume, that you have to choose some unpopular and niche technology/framework.

But I gave you examples of one of the most popular frameworks - Qt and Xamarin...

> But instead of testing on 2-3 platforms, you have to reimplement the same feature several times if you want to have feature parity.

Yes, I agree with that. Unfortunately, this is our way of working now. I implement a feature for Android, and somebody else implements it for iOS. I don't really like but there is nothing we can do about that mostly because of my first point. Plus I don't believe it's worth it (just IMO).

>I think you should check Google Earth. There is an interesting story about reimplementing their GUI to Flutter https://medium.com/flutter/extreme-ui-adaptability-in-flutter-how-google-earth-supports-every-use-case-on-earth-6db4661e7a17
>
>You could also check an app with complex logic and native rendering called PUBG, which incorporated flutter as a part of their frontend to avoid rewriting UI many times [link to short dev story](https://www.youtube.com/watch?v=03ETVuWZiBc)
>
>Are these 2 examples complex and real enough?

Yes. Especially PUBG. It's a really good example. Thanks for that.

The only chance that this topic would be seriously considered and possibly implemented someday in OM is when somebody (and I'm pointing at you @pktiuk because it was your idea, nothing personal 😄) will prepare a PR with the implementation of this feature. Not the whole app, ofc, just POC to show that it is indeed possible and it is worth it.
And with this PR you then go to our @organicmaps/cofounders and try to convince them that it's a good decision and they really need it. And this would be really a hard thing to do. 
Until then there is nothing to discuss or think about. I don't believe in this idea (honestly, after I saw PUBG I started to like it a bit more) and (I think) they don't believe in this idea either.

That's why I marked this issue as "Won't fix".

**Comment by pktiuk** [comment]
> The only chance that this topic would be seriously considered and possibly implemented someday in OM is when somebody (and I'm pointing at you @pktiuk because it was your idea, nothing personal 😄) will prepare a PR with the implementation of this feature. Not the whole app, ofc, just POC to show that it is indeed possible and it is worth it.

This seems to be a good idea. :)     
Maybe I will try to implement something in my spare free time. But I don't guarantee anything.
